### ISSUE TITLE
Speed up basis matrix column checking

### ISSUE THREAD

**Issue by jellis18** [nanograv/enterprise#122]
Currently we always go through every column of every basis matrix to check if any columns are identical. This is very useful since it allows us to use an arbitrary number (and value) of frequencies for different red noise processes without using more columns in the overall basis matrix than we need. For example, if we are using the same 30 frequencies for red noise and  GWB then the overall basis matrix should only have 60 values (sine and cosine double the value) and not 120. 

This all works fine but the method that we are currently using is very brute force and **very** slow. It takes several minutes for the newer data sets. This is not terrible since we only have to do it once but it is still a bit annoying.

Does anyone have any thoughts on how to speed this up?

**Comment by vallis** [comment]
Assuming that the columns would be exactly identical (not just very close), some form of hashing could speed up the comparison. Is that really what's most expensive though, or is it the repeated computation of the same column?

**Comment by jellis18** [comment]
Its the repeated computation. One way to avoid this (which brings up another issue) is to only do this for the Fourier basis components, since those are the only ones that would need this functionality anyway.

The problem with this is that currently there is no enforced naming of signals. All signals have a `signal_type` and `signal_name` but there is nothing that requires that these are set of that these are different for each signal or that they even bare any resemblance to the signal itself.

Maybe we should nave a name required for each signal that would then be prepended with the pulsar name when initialized...

**Comment by jellis18** [comment]
> The way we're doing them now, the comparisons seem very wasteful. Are they not a factor in timings

Yes, maybe we are misunderstanding each other. They are a huge factor. The problem is that the overall basis matrix can be up to 40,000 x 1,000 and we are doing comparisons of every row with every other so it goes like 1e6 comparisons of 40,000 dimensional vectors. The large number of columns comes from the timing model and/or ECORR if done in the basis way. If we only had the fourier basis then it wold still be inefficient but we would only need to compare a few 100 columns at most

**Comment by vallis** [comment]
Ah then a hash is the way to go. Sounds fun, I will try

**Comment by vallis** [comment]
Something like this should work...

```python
        def _combine_basis_columns(self, signals):
            """Given a set of Signal objects, each of which may return an
            Fmat (through get_basis()), return a dict (indexed by signal)
            of integer arrays that map individual Fmat columns to the
            combined Fmat.

            Note: The Fmat returned here is simply meant to initialize the
            matrix to save computations when calling `get_basis` later.
            """

            idx, Fmatlist, hashlist = {}, [], []
            cc = 0
            for signal in signals:
                Fmat = signal.get_basis()

                if Fmat is not None and not signal.basis_params:
                    idx[signal] = []

                    for i, column in enumerate(Fmat.T):
                        colhash = hash(column.tostring())
                        try:
                            j = hashlist.index(colhash)
                            idx[signal].append(j)
                        except ValueError:
                            idx[signal].append(cc)
                            Fmatlist.append(column)
                            hashlist.append(colhash)
                            cc += 1
                elif Fmat is not None and signal.basis_params:
                    nf = Fmat.shape[1]
                    idx[signal] = list(np.arange(cc, cc+nf))
                    cc += nf

            ncol = len(np.unique(sum(idx.values(), [])))
            nrow = len(Fmatlist[0])
            return idx, np.zeros((nrow, ncol))
```

This runs, but currently fails the test, so I'm probably doing something silly. But I need to leave the office right now, so I can't diagnose it...

**Comment by jellis18** [comment]
@vallis, this is good. For some reason the columns are not completely identical which is why I was using `np.allclose` before. If I convert the columns to single precision `colhash = hash(column.astype('float32').tostring())` then I believe this will work. I'll make a branch and run some more tests.

**Comment by jellis18** [comment]
This is a HUGE speedup! The results below are for the 12.5 year 1713 data set

new: 10 loops, best of 3: 198 ms per loop
old: 1 loop, best of 3: 3min 12s per loop

**Comment by scottransom** [comment]
Heh.  That's pretty awesome.  And it is giving the same answers?  If so, optimization via correct algorithm choice FTW!

**Comment by jellis18** [comment]
So this works and passes the tests but I had to convert the columns to 8-character strings before hashing.  This is a little worrisome though...

**Comment by vallis** [comment]
Hmm... but doing that prints out the decimal values to 5/6 digits... Worrisome indeed. The 'float32' solution did not get rid of roundoff discrepancies?

**Comment by jellis18** [comment]
It did for the most part but I ran into problems in the unit tests, although it only happened for PINT. I'll upload some figures in a few minutes

**Comment by vallis** [comment]
Maybe we can get to the bottom of why the bases are slightly different. Sines and cosines should be deterministic.

**Comment by jellis18** [comment]
OK, it looks like it is somehow related to `PINT`. I run this code (from the tests directory of enterprise)
```python
psr = Pulsar('data/B1855+09_NANOGrav_9yv1.gls.par', 'data/B1855+09_NANOGrav_9yv1.tim', 
             timing_package='pint', ephem='DE430')

pl = utils.powerlaw(log10_A=parameter.Uniform(-18,-12),
                    gamma=parameter.Uniform(1,7))
cpl = utils.powerlaw(log10_A=parameter.Uniform(-18,-12)('log10_Agw'),
                     gamma=parameter.Uniform(1,7)('gamma_gw'))

# parameters
log10_A, gamma = -14.5, 4.33
log10_Ac, gammac = -15.5, 1.33
params = {'B1855+09_log10_A': log10_A,
          'B1855+09_gamma': gamma,
          'log10_Agw': log10_Ac,
          'gamma_gw': gammac}

Tmax = psr.toas.max() - psr.toas.min()
nf1 = 30
nf2 = 20
rn = gp_signals.FourierBasisGP(spectrum=pl, components=nf1, Tspan=Tmax)
crn = gp_signals.FourierBasisGP(spectrum=cpl, components=nf2, Tspan=Tmax)
tm = gp_signals.TimingModel()
s = rn + crn + tm
rnm = s(psr)

F1, f1 = utils.createfourierdesignmatrix_red(
    psr.toas, nmodes=nf1, Tspan=Tmax)
F2, f2 = utils.createfourierdesignmatrix_red(
    psr.toas, nmodes=nf2, Tspan=Tmax)

plt.figure(figsize=(20,16))
for ii in range(40):
    plt.subplot(8,5,ii+1)
    plt.plot(psr.toas, F1[:,ii].astype('float32')-F2[:,ii].astype('float32'))
    plt.title(f1[ii]-f2[ii])
plt.tight_layout()
plt.savefig('diff_32_tempo2.png', dpi=200, bbox_inches='tight')

plt.figure(figsize=(20,16))
for ii in range(40):
    plt.subplot(8,5,ii+1)
    plt.plot(psr.toas, F1[:,ii]-F2[:,ii])
    plt.title(f1[ii]-f2[ii])
plt.tight_layout()
plt.savefig('diff_64_tempo2.png', dpi=200, bbox_inches='tight')
```
 And I get these outputs with `PINT`:

Using `float32`
![diff_32](https://user-images.githubusercontent.com/3458354/31243218-9270bf32-a9d7-11e7-9a70-daf81ce626ba.png)

Using `float64`
![diff_64](https://user-images.githubusercontent.com/3458354/31243548-a9c36522-a9d7-11e7-9a3c-f80276e54515.png)

but with `tempo2` I just get 0 difference. I will look into this some more and try to figure out what is going on but `float32` should be fine.

**Comment by vallis** [comment]
float32 is still different in one case though... why are the frequencies different? It's not like Tmax is changing...

**Comment by jellis18** [comment]
I don't know, it is some weird black magic...

**Comment by vallis** [comment]
Ah, it's `np.linspace`:

```
>>> np.any(np.linspace(1/Tmax, 30/Tmax, 30)[:20] - np.linspace(1/Tmax, 20/Tmax, 20) != 0)
True
```

We need something like `(1/Tmax) * np.arange(1, nmodes+1)` (which however is not general enough).

**Comment by jellis18** [comment]
Yea but this doesn't happen when I used `tempo2`. Maybe `Tmax` is just different enough that it is only a problem in certain cases?

**Comment by vallis** [comment]
It would have to be something like that. Note also that toas are numpy.longdouble with libstempo; I don't know about PINT. Does `Pulsar` cast them down to float64 in both cases?

**Comment by jellis18** [comment]
Yes

**Comment by vallis** [comment]
OK, so back to your hypothesis. We can still rewrite `createfourierdesignmatrix_red` so that partially overlapping frequency arrays have identically zero differences.

**Comment by vallis** [comment]
In my latest commit to #119 (my master), I do

```
    # define sampling frequencies
    if fmin is None and fmax is None and not logf:
        # make sure partially overlapping sets of modes
        # have identical frequencies
        f = 1.0 * np.arange(1, nmodes + 1) / T
    else:
        # more general case

        if fmin is None:
            fmin = 1 / T

        if fmax is None:
            fmax = nmodes / T

        if logf:
            f = np.logspace(np.log10(fmin), np.log10(fmax), nmodes)
        else:
            f = np.linspace(fmin, fmax, nmodes)
```

If fmin or fmax are prescribed, or for logarithmic sampling, I don't know if we can achieve "perfectly" partially overlapping mode sets.

**Comment by jellis18** [comment]
Ok this was fixed in PR #126
